File: cda-python-components/programmingtheiot/data/DataUtil.py
#####
# 
# This class is part of the Programming the Internet of Things
# project, and is available via the MIT License, which can be
# found in the LICENSE file at the top level of this repository.
# 
# You may find it more helpful to your design to adjust the
# functionality, constants and interfaces (if there are any)
# provided within in order to meet the needs of your specific
# Programming the Internet of Things project.
# 

import json
import logging
from decimal import Decimal
from json import JSONEncoder

from programmingtheiot.data.ActuatorData import ActuatorData
from programmingtheiot.data.SensorData import SensorData
from programmingtheiot.data.SystemPerformanceData import SystemPerformanceData

class DataUtil():
	"""
	Utility class for converting IoT data objects to/from JSON format.
	Supports ActuatorData, SensorData, and SystemPerformanceData conversions.
	"""

	def __init__(self, encodeToUtf8 = False):
		"""
		Constructor for DataUtil.
		
		@param encodeToUtf8 If True, encode JSON to UTF-8 bytes; otherwise, return string
		"""
		self.encodeToUtf8 = encodeToUtf8
		logging.info("Created DataUtil instance.")
	
	def actuatorDataToJson(self, data: ActuatorData = None, useDecForFloat: bool = False):
		"""
		Converts ActuatorData object to JSON string.
		
		@param data The ActuatorData object to convert
		@param useDecForFloat If True, use Decimal for float values
		@return JSON string representation of ActuatorData, or empty string if data is None
		"""
		if not data:
			logging.debug("ActuatorData is null. Returning empty string.")
			return ""
		
		jsonData = self._generateJsonData(obj = data, useDecForFloat = useDecForFloat)
		return jsonData
	
	def sensorDataToJson(self, data: SensorData = None, useDecForFloat: bool = False):
		"""
		Converts SensorData object to JSON string.
		
		@param data The SensorData object to convert
		@param useDecForFloat If True, use Decimal for float values
		@return JSON string representation of SensorData, or empty string if data is None
		"""
		if not data:
			logging.debug("SensorData is null. Returning empty string.")
			return ""
		
		jsonData = self._generateJsonData(obj = data, useDecForFloat = useDecForFloat)
		return jsonData

	def systemPerformanceDataToJson(self, data: SystemPerformanceData = None, useDecForFloat: bool = False):
		"""
		Converts SystemPerformanceData object to JSON string.
		
		@param data The SystemPerformanceData object to convert
		@param useDecForFloat If True, use Decimal for float values
		@return JSON string representation of SystemPerformanceData, or empty string if data is None
		"""
		if not data:
			logging.debug("SystemPerformanceData is null. Returning empty string.")
			return ""
		
		jsonData = self._generateJsonData(obj = data, useDecForFloat = useDecForFloat)
		return jsonData
	
	def jsonToActuatorData(self, jsonData: str = None, useDecForFloat: bool = False):
		"""
		Converts JSON string to ActuatorData object.
		
		@param jsonData The JSON string to convert
		@param useDecForFloat If True, parse float values as Decimal
		@return ActuatorData object, or None if jsonData is empty/null
		"""
		if not jsonData:
			logging.warning("JSON data is empty or null. Returning null.")
			return None
		
		jsonStruct = self._formatDataAndLoadDictionary(jsonData, useDecForFloat = useDecForFloat)
		ad = ActuatorData()
		self._updateIotData(jsonStruct, ad)
		
		return ad
	
	def jsonToSensorData(self, jsonData: str = None, useDecForFloat: bool = False):
		"""
		Converts JSON string to SensorData object.
		
		@param jsonData The JSON string to convert
		@param useDecForFloat If True, parse float values as Decimal
		@return SensorData object, or None if jsonData is empty/null
		"""
		if not jsonData:
			logging.warning("JSON data is empty or null. Returning null.")
			return None
		
		jsonStruct = self._formatDataAndLoadDictionary(jsonData, useDecForFloat = useDecForFloat)
		sd = SensorData()
		self._updateIotData(jsonStruct, sd)
		
		return sd
	
	def jsonToSystemPerformanceData(self, jsonData: str = None, useDecForFloat: bool = False):
		"""
		Converts JSON string to SystemPerformanceData object.
		
		@param jsonData The JSON string to convert
		@param useDecForFloat If True, parse float values as Decimal
		@return SystemPerformanceData object, or None if jsonData is empty/null
		"""
		if not jsonData:
			logging.warning("JSON data is empty or null. Returning null.")
			return None
		
		jsonStruct = self._formatDataAndLoadDictionary(jsonData, useDecForFloat = useDecForFloat)
		spd = SystemPerformanceData()
		self._updateIotData(jsonStruct, spd)
		
		return spd
	
	def _formatDataAndLoadDictionary(self, jsonData: str, useDecForFloat: bool = False) -> dict:
		"""
		Formats JSON string and converts it to a Python dictionary.
		Replaces single quotes with double quotes and Python boolean strings with JSON boolean strings.
		
		@param jsonData The JSON string to format and parse
		@param useDecForFloat If True, parse float values as Decimal
		@return Dictionary representation of the JSON data
		"""
		# Replace Python-style strings with JSON-compatible strings
		jsonData = jsonData.replace("\'", "\"").replace('False', 'false').replace('True', 'true')
		
		jsonStruct = None
		
		if useDecForFloat:
			jsonStruct = json.loads(jsonData, parse_float = Decimal)
		else:
			jsonStruct = json.loads(jsonData)
		
		return jsonStruct
	
	def _generateJsonData(self, obj, useDecForFloat: bool = False) -> str:
		"""
		Generates JSON string from a Python object using JsonDataEncoder.
		
		@param obj The object to convert to JSON
		@param useDecForFloat If True, use Decimal for float values (not yet implemented)
		@return JSON string representation of the object
		"""
		jsonData = None
		
		if self.encodeToUtf8:
			# Encode to UTF-8 bytes
			jsonData = json.dumps(obj, cls = JsonDataEncoder).encode('utf8')
		else:
			# Return formatted JSON string with indentation
			jsonData = json.dumps(obj, cls = JsonDataEncoder, indent = 4)
		
		if jsonData:
			# Replace Python-style strings with JSON-compatible strings
			jsonData = jsonData.replace("\'", "\"").replace('False', 'false').replace('True', 'true')
		
		return jsonData
	
	def _updateIotData(self, jsonStruct, obj):
		"""
		Updates an IoT data object's attributes from a JSON structure (dictionary).
		Only updates attributes that exist in the object.
		
		@param jsonStruct Dictionary containing the JSON data
		@param obj The IoT data object to update (ActuatorData, SensorData, or SystemPerformanceData)
		"""
		# Get the object's attributes as a dictionary
		varStruct = vars(obj)
		
		# Iterate through JSON keys and update object attributes
		for key in jsonStruct:
			if key in varStruct:
				setattr(obj, key, jsonStruct[key])
				logging.debug("JSON data contains key mappable to object: %s", key)
			else:
				logging.warning("JSON data contains key not mappable to object: %s", key)

	
class JsonDataEncoder(JSONEncoder):
	"""
	Convenience class to facilitate JSON encoding of an object that
	can be converted to a dict.
	"""
	def default(self, o):
		"""
		Override the default method to convert objects to dictionaries.
		
		@param o The object to encode
		@return Dictionary representation of the object using __dict__
		"""
		return o.__dict__

File: gda-java-components/src/main/java/programmingtheiot/data/DataUtil.java
/**
 * This class is part of the Programming the Internet of Things
 * project, and is available via the MIT License, which can be
 * found in the LICENSE file at the top level of this repository.
 * 
 * You may find it more helpful to your design to adjust the
 * functionality, constants and interfaces (if there are any)
 * provided within in order to meet the needs of your specific
 * Programming the Internet of Things project.
 */

package programmingtheiot.data;

import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.google.gson.Gson;

import programmingtheiot.common.ConfigConst;

/**
 * Singleton utility class for converting IoT data objects to/from JSON format.
 * Uses Google's Gson library for JSON serialization and deserialization.
 * 
 * Supports conversion for:
 * - ActuatorData
 * - SensorData
 * - SystemPerformanceData
 * - SystemStateData (optional)
 */
public class DataUtil
{
	// static
	
	private static final Logger _Logger =
		Logger.getLogger(DataUtil.class.getName());
	
	private static final DataUtil _Instance = new DataUtil();

	/**
	 * Returns the Singleton instance of this class.
	 * 
	 * @return DataUtil The singleton instance
	 */
	public static final DataUtil getInstance()
	{
		return _Instance;
	}
	
	
	// private var's
	
	
	// constructors
	
	/**
	 * Default (private) constructor.
	 * Initializes the DataUtil instance.
	 */
	private DataUtil()
	{
		super();
		_Logger.info("DataUtil instance created.");
	}
	
	
	// public methods
	
	/**
	 * Converts an ActuatorData object to a JSON string.
	 * 
	 * @param actuatorData The ActuatorData instance to convert
	 * @return String JSON representation of the ActuatorData, or null if input is null
	 */
	public String actuatorDataToJson(ActuatorData actuatorData)
	{
		String jsonData = null;
		
		if (actuatorData != null) {
			_Logger.fine("Converting ActuatorData to JSON: " + actuatorData.getName());
			
			Gson gson = new Gson();
			jsonData = gson.toJson(actuatorData);
			
			_Logger.fine("ActuatorData converted to JSON: " + jsonData);
		} else {
			_Logger.warning("ActuatorData is null. Returning null JSON string.");
		}
		
		return jsonData;
	}
	
	/**
	 * Converts a SensorData object to a JSON string.
	 * 
	 * @param sensorData The SensorData instance to convert
	 * @return String JSON representation of the SensorData, or null if input is null
	 */
	public String sensorDataToJson(SensorData sensorData)
	{
		String jsonData = null;
		
		if (sensorData != null) {
			_Logger.fine("Converting SensorData to JSON: " + sensorData.getName());
			
			Gson gson = new Gson();
			jsonData = gson.toJson(sensorData);
			
			_Logger.fine("SensorData converted to JSON: " + jsonData);
		} else {
			_Logger.warning("SensorData is null. Returning null JSON string.");
		}
		
		return jsonData;
	}
	
	/**
	 * Converts a SystemPerformanceData object to a JSON string.
	 * 
	 * @param sysPerfData The SystemPerformanceData instance to convert
	 * @return String JSON representation of the SystemPerformanceData, or null if input is null
	 */
	public String systemPerformanceDataToJson(SystemPerformanceData sysPerfData)
	{
		String jsonData = null;
		
		if (sysPerfData != null) {
			_Logger.fine("Converting SystemPerformanceData to JSON: " + sysPerfData.getName());
			
			Gson gson = new Gson();
			jsonData = gson.toJson(sysPerfData);
			
			_Logger.fine("SystemPerformanceData converted to JSON: " + jsonData);
		} else {
			_Logger.warning("SystemPerformanceData is null. Returning null JSON string.");
		}
		
		return jsonData;
	}
	
	/**
	 * Converts a SystemStateData object to a JSON string.
	 * OPTIONAL: Only implement if you have created SystemStateData class.
	 * 
	 * @param sysStateData The SystemStateData instance to convert
	 * @return String JSON representation of the SystemStateData, or null if input is null
	 */
	public String systemStateDataToJson(SystemStateData sysStateData)
	{
		String jsonData = null;
		
		if (sysStateData != null) {
			_Logger.fine("Converting SystemStateData to JSON: " + sysStateData.getName());
			
			Gson gson = new Gson();
			jsonData = gson.toJson(sysStateData);
			
			_Logger.fine("SystemStateData converted to JSON: " + jsonData);
		} else {
			_Logger.warning("SystemStateData is null. Returning null JSON string.");
		}
		
		return jsonData;
	}
	
	/**
	 * Converts a JSON string to an ActuatorData object.
	 * 
	 * @param jsonData The JSON string to convert
	 * @return ActuatorData The deserialized ActuatorData instance, or null if input is null/empty
	 */
	public ActuatorData jsonToActuatorData(String jsonData)
	{
		ActuatorData data = null;
		
		if (jsonData != null && jsonData.trim().length() > 0) {
			_Logger.fine("Converting JSON to ActuatorData: " + jsonData);
			
			Gson gson = new Gson();
			data = gson.fromJson(jsonData, ActuatorData.class);
			
			_Logger.fine("JSON converted to ActuatorData: " + data.getName());
		} else {
			_Logger.warning("JSON data is null or empty. Returning null ActuatorData.");
		}
		
		return data;
	}
	
	/**
	 * Converts a JSON string to a SensorData object.
	 * 
	 * @param jsonData The JSON string to convert
	 * @return SensorData The deserialized SensorData instance, or null if input is null/empty
	 */
	public SensorData jsonToSensorData(String jsonData)
	{
		SensorData data = null;
		
		if (jsonData != null && jsonData.trim().length() > 0) {
			_Logger.fine("Converting JSON to SensorData: " + jsonData);
			
			Gson gson = new Gson();
			data = gson.fromJson(jsonData, SensorData.class);
			
			_Logger.fine("JSON converted to SensorData: " + data.getName());
		} else {
			_Logger.warning("JSON data is null or empty. Returning null SensorData.");
		}
		
		return data;
	}
	
	/**
	 * Converts a JSON string to a SystemPerformanceData object.
	 * 
	 * @param jsonData The JSON string to convert
	 * @return SystemPerformanceData The deserialized SystemPerformanceData instance, or null if input is null/empty
	 */
	public SystemPerformanceData jsonToSystemPerformanceData(String jsonData)
	{
		SystemPerformanceData data = null;
		
		if (jsonData != null && jsonData.trim().length() > 0) {
			_Logger.fine("Converting JSON to SystemPerformanceData: " + jsonData);
			
			Gson gson = new Gson();
			data = gson.fromJson(jsonData, SystemPerformanceData.class);
			
			_Logger.fine("JSON converted to SystemPerformanceData: " + data.getName());
		} else {
			_Logger.warning("JSON data is null or empty. Returning null SystemPerformanceData.");
		}
		
		return data;
	}
	
	/**
	 * Converts a JSON string to a SystemStateData object.
	 * OPTIONAL: Only implement if you have created SystemStateData class.
	 * 
	 * @param jsonData The JSON string to convert
	 * @return SystemStateData The deserialized SystemStateData instance, or null if input is null/empty
	 */
	public SystemStateData jsonToSystemStateData(String jsonData)
	{
		SystemStateData data = null;
		
		if (jsonData != null && jsonData.trim().length() > 0) {
			_Logger.fine("Converting JSON to SystemStateData: " + jsonData);
			
			Gson gson = new Gson();
			data = gson.fromJson(jsonData, SystemStateData.class);
			
			_Logger.fine("JSON converted to SystemStateData: " + data.getName());
		} else {
			_Logger.warning("JSON data is null or empty. Returning null SystemStateData.");
		}
		
		return data;
	}
	
}